File: leetcode-java/src/main/java/com/wezhyn/learn/daily/WordSearch.java
package com.wezhyn.learn.daily;

/**
 * @author wezhyn
 * @since 09.13.2020
 */
public class WordSearch {

    private static final int[] OFFSET = new int[]{0, 1, 0, -1, 0};

    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0 || board[0].length == 0 || word.length() == 0) {
            return false;
        }
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] == word.charAt(0) && doExist(board, word, 0, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean doExist(char[][] board, String word, int cur, int i, int j) {
        if (cur == word.length() - 1) {
            return true;
        }
        try {
            board[i][j] = '/';
            for (int k = 0; k < 4; k++) {
                int x = i + OFFSET[k];
                int y = j + OFFSET[k + 1];
                if (x < 0 || y < 0 || x >= board.length || y >= board[x].length) {
                    continue;
                }
                if (board[x][y] == word.charAt(cur + 1) && doExist(board, word, cur + 1, x, y)) {
                    return true;
                }
            }
            return false;
        } finally {
            board[i][j] = word.charAt(cur);
        }
    }
}

File: src/main/java/com/offer/WordSearch.kt
package com.offer

/**
 *
 * @author wezhyn
 * @since 06.27.2020
 *
 */
class WordSearch {

    companion object {
        val offset = intArrayOf(0, -1, 0, 1, 0)
    }

    fun exist(board: Array<CharArray>, word: String): Boolean {
        for (i in board.indices) {
            for (y in board[i].indices)
                if (board[i][y] == word[0]) {
                    if (backtrack(board, i, y, word, 1)) {
                        return true
                    }
                }
        }
        return false
    }


    private fun backtrack(board: Array<CharArray>, x: Int, y: Int, word: String, index: Int): Boolean {
        val cur = board[x][y]
        board[x][y] = ' '
//        已经将当前字符串全部查找完
        if (index >= word.length) {
            return true
        }
        var result = false
        try {
            for (i in 0..3) {
                val xi = x + offset[i]
                val yi = y + offset[i + 1]
                if (xi < 0 || yi < 0 || xi >= board.size || yi >= board[xi].size) {
                    continue
                }
                if (board[xi][yi] == word[index]) {
                    result = result || backtrack(board, xi, yi, word, index + 1)
                }
            }
            return result
        } finally {
            board[x][y] = cur
        }
    }
}